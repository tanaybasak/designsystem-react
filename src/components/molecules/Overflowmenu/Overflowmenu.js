import React, { useState } from "react";
import PropTypes from "prop-types";
import { prefix } from "../../../settings";
import MenuList from "../../atoms/MenuList";


const Overflowmenu = ({ direction = "left", ...rest }) => {
  const [display, changeDisplay] = useState(false);

  const clickHandler = e => {
    changeDisplay(!display);
  };

  return (
    <section className="hcl-overflow-container">
      <div className={`${prefix}-ellipsis`} onClick={clickHandler} />
      {display && <div
        className={`${prefix}-overflow-menu ${prefix}-overflow-${direction}`}
      ><MenuList items={rest.listItems} />
        <div className={direction === 'left' ? `${prefix}-overflow-caret` : `${prefix}-overflow-caret-right`} />
      </div>}
    </section>
  );
};

Overflowmenu.defaultProps = {
  direction: 'left',
  listItems: null,
  onClick: () => { }
};

Overflowmenu.propTypes = {
  direction: PropTypes.oneOf(['left', 'right']),
  listItems: PropTypes.array.isRequired,
  onClick: PropTypes.func.isRequired
};

export default Overflowmenu;