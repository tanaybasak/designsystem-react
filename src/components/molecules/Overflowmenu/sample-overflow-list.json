[
    {
        "name": "option 1"
            
    },
    {
        "name": "option 2",
        "danger" : true   
    },
    {
        "name": "option 3",
        "separator" : true   
    },
    {
        "name": "option 4",
        "disabled" : true       
    },
    {
        "name": "option 5",
        "link" : "https://google.com"     
    }
]