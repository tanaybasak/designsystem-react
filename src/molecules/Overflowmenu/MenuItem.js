import React from 'react';
import PropTypes from 'prop-types';
import prefix from '../../settings';

const MenuItem = ({
  item,
  onClick,
  disabled,
  danger,
  separator,
  children,
  className,
  ...rest
}) => {
  const classNames = [`${prefix}-overflow-option-btn`];
  if (danger) {
    classNames.push(`${prefix}-overflow-option-dangerbtn`);
  }
  if (disabled) {
    classNames.push(`${prefix}-overflow-option-disablebtn`);
  }
  if (separator) {
    classNames.push(`${prefix}-overflow-separator`);
  }
  if(className){
    classNames.push(className)
  }
  
  return (
    //   {items.map((item, index) => {
    //     const { danger, disabled, separator, link, name, ...rest } = item;
    //     const itemClassNames = [`${prefix}-overflow-option`];
    //     const btnClassNames = [`${prefix}-overflow-option-btn`];

    //     danger && btnClassNames.push(`${prefix}-overflow-option-dangerbtn`);
    //     disabled && btnClassNames.push(`${prefix}-overflow-option-disablebtn`);
    //     separator && btnClassNames.push(`${prefix}-overflow-separator`);

    //     return (
    <li className={`${prefix}-overflow-option`}>
      <button
        className={classNames.join(' ')}
        disabled={disabled}
        onClick={e => {
          if (onClick) {
            onClick(item, e);
          }
        }}
        {...rest}
      >
        {children}
      </button>
    </li>
    //     );
    //   })}
    // </ul>
  );
};

MenuItem.propTypes = {
  item: PropTypes.any,
  onClick: PropTypes.func,
  disabled: PropTypes.bool,
  danger: PropTypes.bool,
  separator: PropTypes.bool,
  children: PropTypes.any
};

MenuItem.defaultProps = {
  item: {},
  onClick: () => {},
  disabled: false,
  danger: false,
  separator: false,
  children: null
};

export default MenuItem;
