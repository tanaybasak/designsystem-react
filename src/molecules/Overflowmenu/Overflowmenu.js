import React, { useState, useEffect, useRef } from 'react';
import PropTypes from 'prop-types';
import MenuList from '../../atoms/MenuList';
import { addListener, removeListeners } from '../../util/eventManager';
import prefix from '../../settings';

let overflowIdRef = 0;
const Overflowmenu = ({
  className,
  direction,
  children,
  ellipsisType,
  onClick,
  listItems,
  ...restProps
}) => {
  const [display, changeDisplay] = useState(false);
  const overflow = useRef(null);
  const [overflowId] = useState(overflowIdRef++);

  useEffect(() => {
    const overflowMenu = overflow.current.children[1];
    if (overflowMenu) {
      const icon = overflow.current.children[0];
      const caret = overflowMenu.children[1];
      let outOfBound = false;
      const parentHeight = (
        overflowMenu.parentElement.offsetHeight +
        8 -
        parseInt(getComputedStyle(icon).marginBottom)
      ).toString();

      overflowMenu.style.display = 'block';
      updateOverflowMenuPos(overflowMenu, icon, caret, outOfBound);
      if (!isInViewport(overflowMenu)) {
        outOfBound = true;
        updateOverflowMenuPos(overflowMenu, icon, caret, outOfBound);
      }
      focusNode(overflowMenu.children[0].children[0]);
      overflowMenu.style.top = parentHeight.concat('px');
    }
  });

  useEffect(() => {
    if (!display) {
      removeListeners('overflow-' + overflowId, 'click');
    } else {
      addListener(
        'overflow-' + overflowId,
        'click',
        e => {
          handleClick(e);
        },
        true
      );
    }
  }, [display]);

  const handleClick = e => {
    if (overflow.current) {
      if (e && overflow.current.contains(e.target)) {
        return;
      }
      changeDisplay(false);
    }
  };

  const clickHandler = () => {
    changeDisplay(!display);
  };

  const isInViewport = elem => {
    const bounding = elem.getBoundingClientRect();
    return (
      bounding.left >= 0 &&
      bounding.right <=
        (window.innerWidth || document.documentElement.clientWidth)
    );
  };

  const updateOverflowMenuPos = (overflowMenu, icon, caret, outOfBound) => {
    let caretPosition;
    if (overflowMenu.classList.contains(`${prefix}-overflow-right`)) {
      if (outOfBound) {
        caretPosition = (icon.offsetWidth / 2 - 18).toString();
        overflowMenu.style.left = null;
        overflowMenu.style.right = caretPosition.concat('px');
        caret.style.left = '10rem';
      } else {
        caretPosition = (icon.offsetWidth / 2 - 22).toString();
        overflowMenu.style.left = caretPosition.concat('px');
        overflowMenu.style.right = null;
        caret.style.left = null;
      }
    } else if (overflowMenu.classList.contains(`${prefix}-overflow-left`)) {
      if (outOfBound) {
        caretPosition = (icon.offsetWidth / 2 - 22).toString();
        overflowMenu.style.right = null;
        overflowMenu.style.left = caretPosition.concat('px');
        caret.style.left = '1rem';
      } else {
        caretPosition = (icon.offsetWidth / 2 - 18).toString();
        overflowMenu.style.right = caretPosition.concat('px');
        overflowMenu.style.left = null;
        caret.style.left = null;
      }
    }
  };

  const focusNode = node => {
    if (node.classList.contains(`${prefix}-overflow-option`)) {
      node.children[0].focus();
    }
  };

  const keyDownOnOverflow = e => {
    const key = e.which || e.keyCode;
    const listItem = e.target.parentElement;
    switch (key) {
      case 40: {
        if (!listItem.nextElementSibling) {
          focusNode(listItem.parentElement.firstElementChild);
        } else if (listItem.nextElementSibling.children[0].disabled === true) {
          focusNode(listItem.nextElementSibling.nextElementSibling);
        } else {
          focusNode(listItem.nextElementSibling);
        }
        e.preventDefault();
        break;
      }
      case 38: {
        if (!listItem.previousElementSibling) {
          focusNode(listItem.parentElement.lastElementChild);
        } else if (
          listItem.previousElementSibling.children[0].disabled === true
        ) {
          focusNode(listItem.previousElementSibling.previousElementSibling);
        } else {
          focusNode(listItem.previousElementSibling);
        }
        e.preventDefault();
        break;
      }
      case 13: {
        e.preventDefault();
        e.target.click();
        break;
      }
      default:
        break;
    }
  };

  const toggleOverflow = e => {
    if (e.key === 'Enter') {
      e.preventDefault();
      overflow.current.children[0].click();
    }
  };

  let element = null;
  if (typeof children !== 'string') {
    element = React.Children.map(children, child => {
      return React.cloneElement(child, {
        tabIndex: '0',
        onClick: clickHandler,
        onKeyPress: toggleOverflow
      });
    });
  }

  const classnames = `${prefix}-overflow-container ${className}`.trim();

  return (
    <section className={classnames} {...restProps} ref={overflow}>
      {typeof children === 'string' ? (
        <span tabIndex="0" onKeyPress={toggleOverflow} onClick={clickHandler}>
          {children}
        </span>
      ) : children === null ? (
        <span
          tabIndex="0"
          onKeyPress={toggleOverflow}
          className={`${prefix}-ellipsis${
            ellipsisType === 'horizontal' ? ' horizontal-ellipsis' : ''
          }`}
          onClick={clickHandler}
        >
          {children}
        </span>
      ) : (
        element
      )}

      {display && Array.isArray(listItems) && listItems.length > 0 && (
        <div
          onKeyDown={keyDownOnOverflow}
          style={{ display: 'none' }}
          className={`${prefix}-overflow-menu ${prefix}-overflow-${direction}`}
        >
          <MenuList
            items={listItems}
            onSelect={event => {
              changeDisplay(false);
              onClick(event);
            }}
          />
          <div
            className={`${prefix}-overflow-caret${
              direction === 'left' ? '' : '-right'
            }`}
          />
        </div>
      )}
    </section>
  );
};

Overflowmenu.defaultProps = {
  direction: 'left',
  listItems: null,
  children: null,
  ellipsisType: 'vertical',
  onClick: () => {},
  className: ''
};

Overflowmenu.propTypes = {
  /** Left: To open overflow menu in left direction. 
  Right: To open overflow menu in right direction  */
  direction: PropTypes.oneOf(['left', 'right']),

  children: PropTypes.node,

  /** List of the item and associated actions, which should be part of menu. */
  listItems: PropTypes.array.isRequired,

  /** Horizontal: To make ellipsis horizontal. 
  Vertical: To make ellipsis vertical.  */
  ellipsisType: PropTypes.oneOf(['vertical', 'horizontal']),

  /** Callback function on selecting item*/
  onClick: PropTypes.func.isRequired,

  /** Class/clasess will be applied on the parent div of OverflowMenu */
  className: PropTypes.string
};

export default Overflowmenu;
