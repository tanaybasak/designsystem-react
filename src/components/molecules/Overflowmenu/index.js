import Overflowmenu from './Overflowmenu';
export default Overflowmenu;