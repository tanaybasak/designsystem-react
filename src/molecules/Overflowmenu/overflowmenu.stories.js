import React from 'react';
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import { select, text, boolean, object } from '@storybook/addon-knobs';
/*
 *@ModuleStart
 */
import Overflowmenu from './Overflowmenu';
/*
 *@ModuleEnd
 */
import '../../story.css';
import 'patron-css/dist/patron-style.css';

const typeOptions = {
  left: 'left',
  right: 'right'
};

const ellipsisType = {
  Vertical: 'vertical',
  Horizontal: 'horizontal'
};

const listItems = [
  {
    name: 'option 1'
  },
  {
    danger: true,
    name: 'option 2'
  },
  {
    name: 'option 3',
    separator: true
  },
  {
    disabled: true,
    name: 'option 4'
  },
  {
    link: 'https://google.com',
    name: 'option 5'
  }
];

storiesOf('Overflowmenu', module).add(
  'basic',
  () => (
    <Overflowmenu
      direction={select('Direction', typeOptions, 'right')}
      ellipsisType={select('ellipsisType', ellipsisType, 'vertical')}
      listItems={object('List Items', listItems)}
      onClick={action('Overflow-Click')}
    />
  ),
  {
    info: {
      text: 'Description About Overflowmenu Component'
    }
  }
);
