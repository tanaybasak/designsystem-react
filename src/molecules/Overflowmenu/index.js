import Overflowmenu from './Overflowmenu';
import MenuItem from './MenuItem';

export { Overflowmenu, MenuItem };
