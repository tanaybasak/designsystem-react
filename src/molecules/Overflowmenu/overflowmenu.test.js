import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import prefix from '../../settings';
import overflowlist from '../../molecules/Overflowmenu/sample-overflow-list.json';
import Overflowmenu from './Overflowmenu';

describe('<Overflowmenu> component', () => {
  it('snapshots/renders Overflowmenu correctly', () => {
    const tree = renderer
      .create(<Overflowmenu listItems={overflowlist} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('triggers click event in Overflowmenu', () => {
    const mockonclick = jest.fn();
    const overflowmenu = mount(
      <Overflowmenu listItems={overflowlist} onClick={mockonclick} />
    );
    overflowmenu.find(`.${prefix}-ellipsis`).simulate('click');
    expect(overflowmenu.find(`.${prefix}-overflow-menu`).exists()).toBeTruthy();
  });

  it('select option on click', () => {
    const mockonclick = jest.fn();
    const overflowmenu = mount(
      <Overflowmenu listItems={overflowlist} onClick={mockonclick} />
    );
    overflowmenu.find(`.${prefix}-ellipsis`).simulate('click');
    expect(overflowmenu.find(`.${prefix}-overflow-menu`).exists()).toBeTruthy();
    overflowmenu.find('button').at(0).simulate('click');
    expect(overflowmenu.find(`.${prefix}-overflow-menu`).exists()).toBeFalsy();
  });

  it('Navigate Item using Arrow Down', () => {
    const mockonclick = jest.fn();
    const overflowmenu = mount(
      <Overflowmenu listItems={overflowlist} onClick={mockonclick} />
    );
    overflowmenu.find(`.${prefix}-ellipsis`).simulate('click');
    expect(overflowmenu.find(`.${prefix}-overflow-menu`).exists()).toBeTruthy();

    let button = overflowmenu.find(`.${prefix}-overflow-option-item`).at(0);
    expect(button.is(':focus')).toBe(true);

    button.simulate('keydown', { keyCode: 40 });
    button = overflowmenu.find(`.${prefix}-overflow-option-item`).at(1);
    expect(button.is(':focus')).toBe(true);
    expect(button.hasClass(`${prefix}-overflow-option-danger`)).toBeTruthy();

    button.simulate('keydown', { keyCode: 40 });
    button = overflowmenu.find(`.${prefix}-overflow-option-item`).at(2);
    expect(button.is(':focus')).toBe(true);
    expect(button.hasClass(`${prefix}-overflow-option-separator`)).toBeTruthy();

    button.simulate('keydown', { keyCode: 40 });
    button = overflowmenu.find(`.${prefix}-overflow-option-item`).at(3);
    expect(button.is(':focus')).toBe(false);
    expect(button.hasClass(`${prefix}-overflow-option-disabled`)).toBeTruthy();

    button = overflowmenu.find(`.${prefix}-overflow-option-item`).at(4);

    expect(button.is(':focus')).toBe(true);
    expect(overflowmenu.find('a')).toBeTruthy();
  });

  it('Navigate Item using Arrow Up', () => {
    const mockonclick = jest.fn();
    const overflowmenu = mount(
      <Overflowmenu listItems={overflowlist} onClick={mockonclick} />
    );
    overflowmenu.find(`.${prefix}-ellipsis`).simulate('click');
    expect(overflowmenu.find(`.${prefix}-overflow-menu`).exists()).toBeTruthy();

    let button = overflowmenu.find(`.${prefix}-overflow-option-item`).at(0);
    expect(button.is(':focus')).toBe(true);

    button.simulate('focus');
    button.simulate('keydown', { keyCode: 38 });
    button = overflowmenu.find(`.${prefix}-overflow-option-item`).at(4);

    expect(button.is(':focus')).toBe(true);
    expect(overflowmenu.find('a')).toBeTruthy();

    button.simulate('focus');
    button.simulate('keydown', { keyCode: 38 });
    button = overflowmenu.find(`.${prefix}-overflow-option-item`).at(3);

    expect(button.is(':focus')).toBe(false);
    expect(button.hasClass(`${prefix}-overflow-option-disabled`)).toBeTruthy();

    button.simulate('focus');
    button.simulate('keydown', { keyCode: 38 });
    button = overflowmenu.find(`.${prefix}-overflow-option-item`).at(2);

    expect(button.is(':focus')).toBe(true);
    expect(button.hasClass(`${prefix}-overflow-option-separator`)).toBeTruthy();

    button.simulate('focus');
    button.simulate('keydown', { keyCode: 38 });
    button = overflowmenu.find(`.${prefix}-overflow-option-item`).at(1);

    expect(button.is(':focus')).toBe(true);
    expect(button.hasClass(`${prefix}-overflow-option-danger`)).toBeTruthy();
  });
});
