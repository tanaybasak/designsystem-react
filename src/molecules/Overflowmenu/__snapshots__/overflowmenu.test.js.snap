// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Overflowmenu> component snapshots/renders Overflowmenu correctly 1`] = `
<div
  className="hcl-overlay-wrapper"
>
  <button
    aria-label="Overflow Menu"
    className="
            hcl-overflow-btn 
            
            
            hcl-ellipsis 
            
            "
    onClick={[Function]}
    role="Overflow Menu"
  />
</div>
`;
